crates/core/tedge_actors/src/converter.rs

```diff
@@ -0,0 +1,203 @@
+use crate::Actor;
+use crate::Builder;
+use crate::DynSender;
+use crate::Message;
+use crate::MessageReceiver;
+use crate::MessageSink;
+use crate::MessageSource;
+use crate::NoConfig;
+use crate::RuntimeError;
+use crate::ServiceConsumer;
+use crate::ServiceProvider;
+use crate::SimpleMessageBox;
+use crate::SimpleMessageBoxBuilder;
+use async_trait::async_trait;
+use std::convert::Infallible;
+
+/// An actor that converts input messages into output messages
+///
+/// ```
+/// # use std::convert::Infallible;
+/// # use crate::tedge_actors::Converter;
+/// # use crate::tedge_actors::ConvertingActor;
+/// # use crate::tedge_actors::RuntimeError;
+/// # use crate::tedge_actors::SimpleMessageBoxBuilder;
+/// # use crate::tedge_actors::ServiceConsumer;
+/// struct Repeater;
+///
+/// impl Converter for Repeater {
+///     type Input = (u8,i32);
+///     type Output = i32;
+///     type Error = Infallible;
+///
+///     fn convert(&mut self, input: &Self::Input) -> Result<Vec<Self::Output>, Self::Error> {
+///         let (n,msg) = *input;
+///         let mut output = vec![];
+///         for _i in 0..n {
+///             output.push(msg);
+///         }
+///         Ok(output)
+///     }
+/// }
+///
+/// # #[tokio::main]
+/// # async fn main() -> Result<(),RuntimeError> {
+/// # use std::time::Duration;
+/// # use tedge_actors::{Actor, Builder, MessageReceiver, MessageSource, NoConfig};
+/// # use tedge_actors::test_helpers::MessageReceiverExt;
+/// let mut actor = ConvertingActor::builder("Repeater", Repeater);
+/// let mut test_box = SimpleMessageBoxBuilder::new("Test", 16).with_connection(&mut actor).build().with_timeout(Duration::from_millis(100));
+/// tokio::spawn(actor.build().run());
+///
+/// test_box.inner.send((3, 42)).await?;
+/// test_box.assert_received([42,42,42]).await;
+///
+/// test_box.inner.send((0, 55)).await?;
+/// test_box.inner.send((2, 1234)).await?;
+/// test_box.assert_received([1234,1234]).await;
+///
+/// assert_eq!(test_box.recv().await, None);
+///
+/// # Ok(())
+/// # }
+/// ```
+pub trait Converter: 'static + Send + Sync {
+    type Input: Message;
+    type Output: Message;
+    type Error: std::error::Error + Send + Sync;
+
+    /// Convert an input message into a vector of output messages
+    fn convert(&mut self, input: &Self::Input) -> Result<Vec<Self::Output>, Self::Error>;
```

Comment:
```suggestion
    fn try_convert(&mut self, input: &Self::Input) -> Result<Vec<Self::Output>, Self::Error>;
```
So that we can have a default impl for a `convert` that maps the errors also into Output.

Reply:
I started with this template in mind - inherited from the `tedge_mapper::Converter()` but finally revert it because not so helpful in the actor itself.